Added multivariate forms of Laplace, studentT and NormalGamma distributions

akash9712 opened the pull request:
Added 3 new types of multivariate distributions. `MultivariateLaplace`
distribution can be called using `Laplace` by using lists/Matrices
instead of real numbers. Multivariate StudentT distribution is called
using `MultivariateT` and not using `StudentT` because of difference in
both type and number of parameters. NormalGamma distribution is called
using `NormalGamma`. Added and modified tests accordingly.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


#### Brief description of what is fixed or changed


#### Other comments

akash9712 committed:
Added multivariate: Laplace, studentT and NormalGamma distributions

Added 3 new types of multivariate distributions. `MultivariateLaplace`
distribution can be called using `Laplace` by using lists/Matrices
instead of real numbers. Multivariate StudentT distribution is called
using `MultivariateT` and not using `StudentT` because of difference in
both type and number of parameters. NormalGamma distribution is called
using `NormalGamma`. Added and modified tests accordingly.

akash9712 commented:
@Upabjojr @jksuom Please have a look.

jksuom commented on sympy/stats/joint_rv_types.py:
```diff
@@ -58,3 +58,152 @@ def marginal_distribution(self, indices, sym):
+        args = Matrix(args)
+        v = 1 - S(k)/2
+        return S(2)/((2*pi)**(S(k)/2)*sqrt(det(sigma))) \
+        *((args.transpose()*sigma_inv*args)[0]/(2 + (mu.transpose()*sigma_inv*mu)[0])) \
```
The returned expression would be more readable (and more efficient) if `args.transpose()` and `mu.transpose()` were precomputed and assigned to local variables. Even `(mu.transpose()*sigma_inv*mu)[0]` could be precomputed.

akash9712 committed:
Make Multivariate Laplace PDF implementation more readable

Used local variables for storing matrices' transpose while calculating
PDF instead of calling `.transpose()` repeatedly.

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -58,3 +58,154 @@ def marginal_distribution(self, indices, sym):
+        _value_check(len(mu) == len(sigma.col(0)),
+            "Size of the mean vector and covariance matrix are incorrect.")
+        #check if covariance matrix is positive definite or not.
+        _value_check(all([i > 0 for i in sigma.eigenvals().keys()]),
```
what about the case in which the eigenvalues cannot be computed?

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -58,3 +58,154 @@ def marginal_distribution(self, indices, sym):
+
+    def pdf(self, *args):
+        from sympy.functions.special.bessel import besselk
+        mu, sigma = Matrix(self.mu), Matrix(self.sigma)
```
`Matrix` not needed here, `self.mu` and `self.sigma` are already supposed to be matrices.

akash9712 replied:
`self.sigma` and `self.mu` may also be of  type `list`, so `Matrix` just makes sure they are transformed into matrices.

Upabjojr replied:
They should be stored as immutable matrices, the transformation should not happen here.

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -58,3 +58,154 @@ def marginal_distribution(self, indices, sym):
+        mu, sigma = Matrix(self.mu), Matrix(self.sigma)
+        mu_T = mu.transpose()
+        k = len(mu)
+        sigma_inv = sigma**(-1)
```
`sigma.inv()` maybe is clearer

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -58,3 +58,154 @@ def marginal_distribution(self, indices, sym):
+        sigma_inv = sigma**(-1)
+        args = Matrix(args)
+        args_T = args.transpose()
+        v = 1 - S(k)/2
```
`S(k)` gets repeated

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -58,3 +58,154 @@ def marginal_distribution(self, indices, sym):
+        v = 1 - S(k)/2
+        return S(2)/((2*pi)**(S(k)/2)*sqrt(det(sigma))) \
+        *((args_T*sigma_inv*args)[0]/(2 + (mu_T*sigma_inv*mu)[0])) \
+        **(S(v)/2)*besselk(v, sqrt((2 + (mu_T*sigma_inv*mu)[0])
```
some repeated expressions could be pre-computed, like ` (mu_T*sigma_inv*mu)[0]`

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -58,3 +58,154 @@ def marginal_distribution(self, indices, sym):
+        _value_check(len(mu) == len(sigma.col(0)),
+            "Size of the location vector and shape matrix are incorrect.")
+        #check if covariance matrix is positive definite or not.
+        _value_check(all([i > 0 for i in sigma.eigenvals().keys()]),
```
eigenval problem?

Upabjojr commented on sympy/stats/joint_rv_types.py:
```diff
@@ -58,3 +58,154 @@ def marginal_distribution(self, indices, sym):
+        _value_check(all([i > 0 for i in sigma.eigenvals().keys()]),
+            "The shape matrix must be positive definite. ")
+
+    def pdf(self, *args):
```
same reasoning

Upabjojr commented:
Do `variance( )`, `expectation( )` and `probability( )` already work with multivariate RV? If yes, it would be nice to have some tests.

akash9712 committed:
Minor changes to joint_rv_types.py

akash9712 commented:
@Upabjojr I have made the changes you mentioned in review comments. `variance( )`, `expectation( )` and `probability( )` do not yet work for these. I will make a seperate PR for them, if that's okay.

Upabjojr commented:
> variance( ), expectation( ) and probability( ) do not yet work for these. I will make a seperate PR for them, if that's okay.

OK
